scvi/module/_mixupvae.py

```diff
@@ -188,22 +189,46 @@ def __init__(
         self.pseudo_bulk = pseudo_bulk
         self.encode_cont_covariates = encode_cont_covariates
         self.mixup_penalty = mixup_penalty
-        # decoder goes from n_latent-dimensional space to n_input-d data
-        n_input_decoder = n_latent + n_continuous_cov
-        use_batch_norm_decoder = use_batch_norm == "decoder" or use_batch_norm == "both"
-        use_layer_norm_decoder = use_layer_norm == "decoder" or use_layer_norm == "both"
-        _extra_decoder_kwargs = extra_decoder_kwargs or {}
-        self.decoder = DecoderSCVI(
-            n_input_decoder,
-            n_input,
-            n_cat_list=None,  # for now, we don't decode categorical variables
+        self.already_printed_warning = False
+
+        # overwrite z_encoder and l_encoder with the right n_input_encoder
+        n_input_encoder = n_input + n_continuous_cov * encode_cont_covariates
+        cat_list = [n_batch] + list([] if n_cats_per_cov is None else n_cats_per_cov)
+        encoder_cat_list = cat_list if encode_covariates else None
+        use_batch_norm_encoder = use_batch_norm == "encoder" or use_batch_norm == "both"
+        use_layer_norm_encoder = use_layer_norm == "encoder" or use_layer_norm == "both"
+        # z encoder goes from the n_input-dimensional data to an n_latent-d
+        # latent space representation
+        _extra_encoder_kwargs = extra_encoder_kwargs or {}
+        self.z_encoder = Encoder(
+            n_input_encoder,
+            n_latent,
+            n_cat_list=encoder_cat_list,
             n_layers=n_layers,
             n_hidden=n_hidden,
+            dropout_rate=dropout_rate,
+            distribution=latent_distribution,
             inject_covariates=deeply_inject_covariates,
-            use_batch_norm=use_batch_norm_decoder,
-            use_layer_norm=use_layer_norm_decoder,
-            scale_activation="softplus" if use_size_factor_key else "softmax",
-            **_extra_decoder_kwargs,
+            use_batch_norm=use_batch_norm_encoder,
+            use_layer_norm=use_layer_norm_encoder,
+            var_activation=var_activation,
+            return_dist=True,
+            **_extra_encoder_kwargs,
+        )
+        # l encoder goes from n_input-dimensional data to 1-d library size
+        self.l_encoder = Encoder(
```

Comment:
Because normally, `n_input_encoder = n_input + n_continuous_cov * encode_covariates`. But here, we want to differentiate between cont covariates (that are still possible to include in our model) and categorical ones (which are not possible).
Therefore we have `encode_covariates = False` pretty much all the time, but a new argument encode_cont_covariates which can be equal to True, with `n_input_encoder = n_input + n_continuous_cov * encode_cont_covariates` now. As this argument changes, we need to re-instantiate the l_encoder and z_encoder so they have the right dimension.

Reply:
"and categorical ones (which are not possible)", we can, it's just simpler for now to start without them. Later we should consider adding the batch index. the problem is how do you define a batch index for your pseudobulk sample. A way could be to remove batch effects first (training with batch index for a few epochs) before introducing the mixup penalty.

Reply:
Okok, well when we define how to do that, we can change these lines :)